typing/jkind.ml

```diff
@@ -2464,11 +2518,89 @@ module Violation = struct
     | Mode
     | Layout
 
+  let report_reason ppf violation =
+    (* Print out per-axis information about why the error occurred. This only happens
+       when modalities are printed because the errors are simple enough when there are no
+       modalities that it makes the error unnecessarily noisy.
+    *)
+    match violation with
+    | Not_a_subjkind (sub, super, reasons) -> (
+      if should_print_with_bounds ()
+      then
+        let disagreeing_axes =
+          (* Collect all the axes that disagree into a set. If none disagree, then it
+             is [None] *)
+          List.fold_left
+            (fun disagreeing_axes_so_far reason ->
+              match
+                (reason : Sub_failure_reason.t), disagreeing_axes_so_far
+              with
+              | Axis_disagreement (Pack axis), Some disagreeing_axes_so_far ->
+                Some (Axis_set.add disagreeing_axes_so_far axis)
+              | Axis_disagreement (Pack axis), None ->
+                Some (Axis_set.singleton axis)
+              | (Layout_disagreement | Constrain_ran_out_of_fuel), _ ->
+                disagreeing_axes_so_far)
+            None reasons
+        in
+        let has_modalities =
+          let jkind_has_modalities jkind =
+            List.exists
+              (fun (_, type_info) ->
+                let axes_ignored_by_modalities =
+                  With_bounds.Type_info.axes_ignored_by_modalities
+                    ~mod_bounds:jkind.jkind.mod_bounds ~type_info
+                in
+                not (Axis_set.is_empty axes_ignored_by_modalities))
+              (With_bounds.to_list jkind.jkind.with_bounds)
+          in
+          jkind_has_modalities sub || jkind_has_modalities super
+        in
+        match disagreeing_axes, has_modalities with
+        | None, _ | _, false -> ()
+        | Some disagreeing_axes, true ->
+          (* CR: @\n is discouraged by the documentation, but @;@; seems to emit one newline
+             and then one space rather than two newlines *)
+          fprintf ppf "@\n@\nThe first mode-crosses less than the second along:";
```

Comment:
I couldn't figure out the "right" way to get a blank line separating the stanzas.

Reply:
I think this is probably fine for now.